helix-core/src/main/java/org/apache/helix/messaging/handling/HelixStateTransitionHandler.java

```diff
@@ -227,56 +172,73 @@ void postHandleMessage() {
         logger.warn("Force CurrentState on Zk to be stateModel's CurrentState. partitionKey: "
             + partitionKey + ", currentState: " + _stateModel.getCurrentState() + ", message: "
             + _message);
         _currentStateDelta.setState(partitionKey, _stateModel.getCurrentState());
-      } else {
-        StateTransitionError error =
-            new StateTransitionError(ErrorType.INTERNAL, ErrorCode.ERROR, exception);
-        if (exception instanceof InterruptedException) {
-          if (_isTimeout) {
-            error = new StateTransitionError(ErrorType.INTERNAL, ErrorCode.TIMEOUT, exception);
-          } else {
-            // State transition interrupted but not caused by timeout. Keep the current
-            // state in this case
-            logger.error(
-                "State transition interrupted but not timeout. Not updating state. Partition : "
-                    + _message.getPartitionName() + " MsgId : " + _message.getMsgId());
-            return;
-          }
-        }
-        _stateModel.rollbackOnError(_message, _notificationContext, error);
-        _currentStateDelta.setState(partitionKey, HelixDefinedState.ERROR.toString());
-        _stateModel.updateState(HelixDefinedState.ERROR.toString());
+        updateZKCurrentState();
+        return;
       }
-    }
+      // Handle timeout interrupt.
+      if ((exception instanceof InterruptedException) && !_isTimeout) {
+        // State transition interrupted but not caused by timeout. Keep the current
+        // state in this case
+        logger.error(
+            "State transition interrupted but not timeout. Not updating state. Partition : "
+                + _message.getPartitionName() + " MsgId : " + _message.getMsgId());
+        return;
+      }
+      // We did early return when Timeout interrupt.
+      StateTransitionError error = new StateTransitionError(ErrorType.INTERNAL,
+          exception instanceof InterruptedException ? ErrorCode.TIMEOUT : ErrorCode.ERROR,
+          exception);
+      _stateModel.rollbackOnError(_message, _notificationContext, error);
+      _currentStateDelta.setState(partitionKey, HelixDefinedState.ERROR.toString());
+      _stateModel.updateState(HelixDefinedState.ERROR.toString());
+    } // NOP if taskResult.isCancelled()
+
+    updateZKCurrentState();
+  }
+
+  // Update the ZK current state of the node
+  private void updateZKCurrentState(){
+    HelixDataAccessor accessor = _manager.getHelixDataAccessor();
+    String partitionKey = _message.getPartitionName();
+    String resource = _message.getResourceName();
+    String sessionId = _message.getTgtSessionId();
+    String instanceName = _manager.getInstanceName();
 
     try {
-      // Update the ZK current state of the node
+      // We did not update _stateModel for DROPPED state, so it won't match _stateModel.
+      if (!_stateModel.getCurrentState().equals(_currentStateDelta.getState(partitionKey))
```

Comment:
nit, this happens when race condition happens, right? Shall we log a warning just for debugging?

Reply:
TFTR. Updated.

Reply:
DROPPED does not count...